### PR TITLE
WIP: rules

### PR DESCRIPTION
**Pull request by skyfrench**
A little progress on #55 for discussion ... particularly how it might change given @TimGuiteDiamond's work on a widget interface.

### CODE REVIEW

**Comment by TimGuiteDiamond** [comment]
This is pretty cool actually! 

I think that it could be quite powerful if wrapped around Widget but yes it would need a bit more intelligence in order to correctly separate out containerStyling vs widgetStyling for example.

**Comment by willrogers** [comment]
Looks nice. Two points:

* `useSubscription` makes it difficult or impossible for you to conform with the hooks rules. Perhaps I could make it take a list of PVs, then you could figure out the ones you want and call from the top-level.
* we should put `useId()` into the top-level `Widget` component when that exists properly, then pass it down as a prop.

**Comment by willrogers** [comment]
Note the ID is per component not per subscription so you could get your ID at the start of `RuleWrapper` and remove one of the warnings. But what I said above applies once we have a Widget component.

I just pushed a `subs` branch that might help with the other warning, it changes `useSubscription` to take an array of PVs.

File: src/components/RuleWrapper/ruleWrapper.tsx [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        });
+        if (latestValue !== undefined && connected) {
+          condition = condition.replace(name, latestValue.getValue());
+        } else valid = false;
+      }
+    }
+    if (valid && condition !== undefined) {
+      let state = evaluate(condition);
+      let styleValue = state ? props.trueState : props.falseState;
+      // use props.prop not hard coded colour
```
Comment (willrogers) [inline src/components/RuleWrapper/ruleWrapper.tsx]: Something like

```
const newProps = { ...props };
newProps[props.prop] = styleValue;
return <Component {...(newProps as P)}></Component>;
```

?

Comment (skyfrench) [reply src/components/RuleWrapper/ruleWrapper.tsx]: That works :)

**Comment by willrogers** [comment]
It now looks to me like we can't call a hook inside a HoC and still obey the 'rules of hooks'.

Are we going to need a refactor? Can we make our HOCs into hooks? I find it hard to think about these things clearly.

**Comment by willrogers** [comment]
You're right I am apparently making the same violation in `connectionWrapper` but I don't get an error. That is weird.

**Comment by willrogers** [comment]
Er... if you change the name from `RuleWrapper` to `ruleWrapper` then two of your errors go away. I'm not sure what to think of this but it's something to do with the wrappers not being components, but returning components.

It still leaves a problem with calling `useSelector` in a loop.

File: src/components/RuleWrapper/ruleWrapper.tsx [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  return (props: RuleProps): JSX.Element => {
+    // eslint-disable-next-line react-hooks/rules-of-hooks
+    const [id] = useId();
+    let condition = props.condition;
+    let valid = true;
+    if (condition === undefined || props.substitutionMap === undefined)
+      valid = false;
+    else {
+      let pvs = [];
+      for (let [pv] of Object.entries(props.substitutionMap)) {
```
Comment (willrogers) [inline src/components/RuleWrapper/ruleWrapper.tsx]: This is picking the key not the value. You could use `Object.values()`

**Comment by willrogers** [comment]
n.b. looks like there are some problems on the `subs` branch that I need to resolve. I'm getting repeated subscriptions and unsubscriptions.

**Comment by willrogers** [comment]
I have pushed another commit to `subs` that resolves the problem I was referring to.

**Comment by willrogers** [comment]
This is now adapted and merged in the `hooks` branch so I'll close this.

**Pull request closed without merging**